### PR TITLE
makeSome refactoring

### PR DESCRIPTION
**Pull request by pablf**
Continuation of #8850

### CODE REVIEW

**Comment by jdegoes** [comment]
@pablf Thanks for re-opening! Can you include /fixes, etc., in the description?

**Comment by pablf** [comment]
@jdegoes the original issues are already closed, should I add something else?

File: core-tests/shared/src/test/scala/zio/autowire/AutoWireSpec.scala [hunk 1]
```diff
@@ -231,11 +250,99 @@ object AutoWireSpec extends ZIOBaseSpec {
...
           }
         )
       )
     )
 
+  def dummy(f: Any*): Boolean = {
+    var l: List[Any] = f.toList
+    l = Nil
+    l.isEmpty
+  }
```
Comment (kyri-petrou) [inline core-tests/shared/src/test/scala/zio/autowire/AutoWireSpec.scala:344]: I'm a little bit confused, what is this testing exactly?

File: core/shared/src/main/scala-2/zio/internal/macros/LayerMacroUtils.scala [hunk 2]
```diff
@@ -161,6 +161,7 @@ private[zio] trait LayerMacroUtils {
     intersectionTypes
       .map(_.dealias)
       .filterNot(_.isAny)
+      .filterNot(t => typeOf[Any] <:< t)
```
Comment (kyri-petrou) [inline core/shared/src/main/scala-2/zio/internal/macros/LayerMacroUtils.scala:163]: Isn't this the same as the above `.filterNot(_.isAny)`?

File: core-tests/shared/src/test/scala/zio/autowire/AutoWireSpec.scala (hunk 1 above)
Comment (pablf) [reply core-tests/shared/src/test/scala/zio/autowire/AutoWireSpec.scala:344]: The test intends to check that the methods before do not create warnings, so it's not useful to test only compilation. Then I got warnings about not using the methods turning into errors. I think I also tried `val _ =` syntax but something didn't work. It's not pretty, maybe there is a better approach...

File: core/shared/src/main/scala-2/zio/internal/macros/LayerMacroUtils.scala (hunk 2 above)
Comment (pablf) [reply core/shared/src/main/scala-2/zio/internal/macros/LayerMacroUtils.scala:163]: I am not sure. I remember that somewhere `ZAny` didn't get filtered and that broke the algorithm and I think it's here.

Comment (pablf) [reply core/shared/src/main/scala-2/zio/internal/macros/LayerMacroUtils.scala:163]: Testing it, `isAny` alone doesn't work, although just ` .filterNot(t => typeOf[Any] <:< t)` works alone

**Review by jdegoes — APPROVED** [review]

**Comment by kyri-petrou** [comment]
Hey @pablf, I just tried to update zio in a codebase at $WORK and I'm getting an exception coming from the macros when using `provideSomeShared` (from zio-test). Do you have any idea why this might be happening?

```
Exception occurred while executing macro expansion.
java.lang.Throwable: This can't happen.
	at zio.internal.macros.Graph.mkNeededKeys$$anonfun$2(Graph.scala:80)
	at zio.internal.macros.Graph.forEach$$anonfun$1(Graph.scala:189)
	at scala.collection.immutable.List.foldRight(List.scala:353)
	at zio.internal.macros.Graph.forEach(Graph.scala:194)
	at zio.internal.macros.Graph.mkNeededKeys(Graph.scala:99)
	at zio.internal.macros.Graph.buildNodes(Graph.scala:27)
	at zio.internal.macros.LayerBuilder.build(LayerBuilder.scala:93)
	at zio.internal.macros.LayerMacroUtils$.constructLayer$$anonfun$1$$anonfun$1(LayerMacroUtils.scala:110)
	at zio.internal.macros.LayerMacroUtils$.constructLayer$$anonfun$1(LayerMacroUtils.scala:110)
	at zio.internal.macros.LayerMacroUtils$.constructLayer$$anonfun$adapted$1(LayerMacroUtils.scala:110)
```

**Comment by pablf** [comment]
Would you mind sharing a similar example or how the layer is being constructed? Also, which scala version is this? It seems like when retrieving a type from `neededKeys` using `get`, it doesn't find it, probably because it was added under a different name. Changing `get` to a comparison made using `keyEquals` or transforming the keys to be retrieved to be the same keys as the one in `neededKeys` should work.

There was a similar issue before that I wasn't able to reproduce so it seems like it was related to some configuration in how Scala treated internally the types. Relevant build details could also be useful. Does this happen in all instances of `provide` methods?

**Comment by guizmaii** [comment]
Reported here too: https://github.com/zio/zio/issues/9145#issue-2486607289

**Pull request merged**